### ISSUE TITLE
ChaCha20 encrypted data not correct?

### ISSUE THREAD

**Issue by romnes** [krzyzanowskim/CryptoSwift#179]
While looking at the ChaCha20 tests, I noticed that expectedHex is never checked. My assumption is that encrypted.toHexString() should equal expectedHex. However, encrypted.toHexString() does not match the expected output in the first test vector (did not run the others). This leads me to believe that the ChaCha20 implementation may not actually be ChaCha20 "compliant", even though decrypting an encrypted message produced by the Swift implementation works.

Given the same key, iv and message I also noticed that I get a different result from a Go and Python implementation using the same key, iv and message. They both produce the same result, but the Swift implementation in CryptoSwift produces a different one:

key: [150, 154, 87, 169, 52, 170, 96, 165, 122, 63, 215, 142, 161, 211, 196, 55, 63, 147, 117, 156, 32, 75, 171, 134, 224, 30, 2, 149, 34, 10, 138, 95]

iv: [80, 83, 45, 77, 115, 103, 48, 53]

message: 64 0 bytes

expected (Go and Python): [216, 16, 193, 127, 240, 7, 33, 238, 146, 7, 74, 133, 22, 23, 81, 141, 142, 210, 154, 250, 57, 23, 52, 2, 181, 122, 115, 134, 138, 111, 199, 164, 174, 140, 45, 14, 213, 3, 175, 29, 207, 62, 230, 11, 33, 21, 44, 37, 175, 167, 195, 37, 108, 148, 8, 132, 230, 0, 52, 193, 168, 177, 165, 150]

actual (Swift): [204, 117, 171, 148, 246, 159, 98, 48, 205, 184, 158, 228, 146, 73, 168, 157, 120, 226, 93, 62, 31, 205, 156, 96, 181, 195, 102, 141, 188, 45, 131, 141, 234, 10, 56, 165, 240, 205, 241, 228, 155, 4, 154, 149, 179, 217, 71, 45, 237, 4, 152, 204, 59, 9, 148, 5, 217, 237, 40, 82, 49, 142, 104, 137]

Thoughts?

**Comment by romnes** [comment]
I found two issues, that when fixed gives me the same output as the Python and Go implementations and work with a third party API that uses ChaCha20 encryption;

First, ChaCha20.swift:56 should be 20, not 10. Alternatively, ChaCha20.swift:66 should decrement with 1, not 2. Either way, the loop should run 10 times according to the RFC.

Second, the convertion [UInt32] to [UInt8] seems to produce bytes in the wrong order. I replaced ChaCha20.swift:74 to use NSData to convert from [UInt32] to [UInt8] which produces the correct order.
